tools/nbs_tools/calibrate_cameras.py

```diff
@@ -0,0 +1,507 @@
+    if not ret:
+        ret, centres = cv2.findCirclesGrid(
+            img, (ROWS, COLS), flags=cv2.CALIB_CB_ASYMMETRIC_GRID | cv2.CALIB_CB_CLUSTERING, blobDetector=detector
+        )
+
+    return {
+        "name": item["camera_name"],
+        "timestamp": item["timestamp"],
+        "bytes_read": item["bytes_read"],
+        "image": img,
+        "blobs": detector.detect(img),
+        "centres": centres if ret else None,
+    }
+
+
+def find_grids(files):
+
+    # Work out what the name of the pickle file will be for caching
+    pickle_path = "{}.pickle".format(
+        "_".join(os.path.basename(f).replace(".nbs", "").replace(".nbz", "").replace(".gz", "") for f in files)
+    )
+
+    # The points from the images that have been gathered
+    if not os.path.isfile(pickle_path):
+
+        # Read the nbs file
+        with multiprocessing.Pool(multiprocessing.cpu_count()) as pool:
+            decoder = Decoder(*files)
+
+            grids = {}
+
+            print("Detecting asymmetric circles grids")
+
+            with tqdm(total=len(decoder), unit="B", unit_scale=True, dynamic_ncols=True) as progress:
+
+                results = []
+
+                # Function that updates the results
+                def update_results(msg):
+
+                    # Update the progress based on the image we are up to
+                    progress.update(msg["bytes_read"])
+
+                    img = msg["image"]
+                    img = cv2.drawKeypoints(
+                        img, msg["blobs"], np.array([]), (0, 0, 255), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS
+                    )
+                    if msg["centres"] is not None:
+                        img = cv2.drawChessboardCorners(img, (ROWS, COLS), msg["centres"], True)
+
+                    cv2.imshow(msg["name"], img)
+                    cv2.waitKey(1)
+
+                    # Add this to the grids list
+                    if msg["name"] not in grids:
+                        grids[msg["name"]] = []
+                    grids[msg["name"]].append(
+                        {
+                            "timestamp": msg["timestamp"],
+                            # Normalise the pixel coordinates to be based from the centre of the image
+                            # And then divide by the width of the image to get a normalised coordinate
+                            "centres": None
+                            if msg["centres"] is None
+                            else (np.array(img.shape[:2][::-1], dtype=np.float) * 0.5 - msg["centres"]) / img.shape[1],
+                        }
+                    )
+
+                for msg in packetise_stream(decoder):
+                    # Add a task to the pool to process
+                    results.append(pool.apply_async(process_frame, (msg,)))
+
+                    # Only buffer 1024 images for each cpu core to avoid running out of memory
+                    if len(results) > 128 * multiprocessing.cpu_count():
```

Comment:
This is why you don't put numbers in comments. It's the number of images that I'm happy to let the CPU queue. Otherwise what can happen is that it loads images to process faster than it processes them until it runs out of ram.